### PR TITLE
Convert to ghc-9.2.1

### PR DESCRIPTION
**Pull request by shayne-fletcher**
This is a complete port of HLint to the ghc-9.2.1 parse tree. All tests pass.

### CODE REVIEW

**Comment by shayne-fletcher** [comment]
Cc @alanz All done mate.

**Comment by ndmitchell** [comment]
Awesome work!

**Comment by ndmitchell** [comment]
GHC 9.2.1 is now out. Thoughts on this?

**Comment by shayne-fletcher** [comment]
> GHC 9.2.1 is now out. Thoughts on this?

Got it. Time to move on it then. Leave it with me for now and I'll keep you posted.

**Comment by alanz** [comment]
For the apply-refact part it will need me to complete my current ghc-exactprint update.  Which is getting close.

**Comment by ndmitchell** [comment]
I'm in no tearing rush - a leisurely upgrade is fine :)

**Comment by shayne-fletcher** [comment]
i rebased this branch on master today (oct 30, 2021).

the tests all pass but this one in 'Extensions.hs':
```
{-# LANGUAGE DeriveGeneric, TypeFamilies #-} \
data family Bar a; data instance Bar Foo = Foo deriving Generic
```
which produces the apparently unexpected output
```
Note: Extension DeriveGeneric is not used
```
(using the 9.2.1 release candidate ghc-lib-parser-9.2.1.20210509.tar.gz that the `stack.yaml` on this branch arranges to use)

this test was introduced only recently in https://github.com/ndmitchell/hlint/pull/1277. @zliu41 as the author of that test might i trouble you to take a look? i removed the test for the moment. 

this puts us in a good place for the next step: upgrade this branch to the ghc-lib-parser and ghc-lib-parser-ex  newly released 9.2.1 official versions.

**Comment by shayne-fletcher** [comment]
fixes issue https://github.com/ndmitchell/hlint/issues/1314

**Comment by zliu41** [comment]
@shayne-fletcher Removing that test for now is good. Once you make cabal build work for this patch, I can look into it - otherwise it's difficult for me to investigate.

**Comment by LeventErkok** [comment]
@shayne-fletcher Maybe it's good to go now for GHC 9.2.1?

**Comment by shayne-fletcher** [comment]
> @shayne-fletcher Maybe it's good to go now for GHC 9.2.1?

yep, it's high priority with us too. don't worry @LeventErkok , we got you! little patience please 😉

**Comment by shayne-fletcher** [comment]
@ndmitchell i believe this is ready to merge. all 813 tests pass locally. please note that minimum build compiler is now 8.10.1.

**Comment by shayne-fletcher** [comment]
p.s. locally my testing was with `refactor-0.9.2.0`

**Comment by shayne-fletcher** [comment]
> @shayne-fletcher Removing that test for now is good. Once you make cabal build work for this patch, I can look into it - otherwise it's difficult for me to investigate.

@zliu41 `cabal v2-build` on this branch works as you'd hope for now!

File: src/Hint/Extensions.hs [hunk 1]
```diff
@@ -515,19 +515,21 @@ addDerives nt _ xs = mempty
 derives :: Located HsModule -> Derives
 derives (L _ m) =  mconcat $ map decl (childrenBi m) ++ map idecl (childrenBi m)
   where
-    idecl :: DataFamInstDecl GhcPs -> Derives
-    idecl (DataFamInstDecl (HsIB _ FamEqn {feqn_rhs=HsDataDefn {dd_ND=dn, dd_derivs=(L _ ds)}})) = g dn ds
+    idecl :: LocatedA (DataFamInstDecl GhcPs) -> Derives
```
Comment (zliu41) [inline src/Hint/Extensions.hs:518]: You should remove the `LocatedA` here. Then the test should pass.

Comment (shayne-fletcher) [reply src/Hint/Extensions.hs]: outstanding @zliu41 , many thanks. i have restored the (now passing) test!

**Comment by ndmitchell** [comment]
Thanks! @shayne-fletcher  - can you add a CI for GHC 9.2 and remove it for GHC 8.8 (see the `.github/workflows`). 

It seems the Windows and Mac CI are both bust installing apply-refact. Any idea why?

I'll take a peek through the code now

File: src/GHC/Util/SrcLoc.hs [hunk 2]
```diff
@@ -25,6 +37,12 @@ newtype SrcSpanD = SrcSpanD SrcSpan
   deriving (Outputable, Eq)
 instance Default SrcSpanD where def = SrcSpanD noSrcSpan
 
+newtype FastStringD = FastStringD FastString
```
Comment (ndmitchell) [inline src/GHC/Util/SrcLoc.hs:40]: Does GHC not provide an Ord for this type? Could we get one added upstream for the future?

**Review by ndmitchell — COMMENTED** [review]
Nice work!

**Comment by shayne-fletcher** [comment]
> Thanks! @shayne-fletcher - can you add a CI for GHC 9.2 and remove it for GHC 8.8 (see the `.github/workflows`).

pushed

> It seems the Windows and Mac CI are both bust installing apply-refact. Any idea why?

no idea about that sorry.

> I'll take a peek through the code now

**Comment by alanz** [comment]
> It seems the Windows and Mac CI are both bust installing apply-refact. Any idea why?

It depends on ghc-exactprint, the release on hackage does not compile with GHC 9.2.1 as released.

I will be making a new release tomorrow,  just need to clean up a few tests

File: src/GHC/Util/SrcLoc.hs (hunk 2 above)
Comment (shayne-fletcher) [reply src/GHC/Util/SrcLoc.hs:40]: @ndmitchell  the `Ord` instance has been removed. see https://gitlab.haskell.org/ghc/ghc/-/commit/4b4fbc58d37d37457144014ef82bdd928de175df. however i overlooked the existence of `LexicalFastString` (see https://hackage.haskell.org/package/ghc-9.2.1/docs/GHC-Data-FastString.html#t:LexicalFastString) which we can use here. pushing a fix for that now.

**Comment by ndmitchell** [comment]
Thanks @alanz - my plan is to hold this PR in limbo a few days, merge nothing else to the repo in the intervening time, and wait for ghc-exactprint, then pull the trigger on this.

**Comment by alanz** [comment]
I just published [ghc-exactprint-1.3.0](https://hackage.haskell.org/package/ghc-exactprint-1.3.0)

**Comment by arjunkathuria** [comment]
hi, if all's good can we merge this ? looking forward : )

File: stack.yaml [hunk 3]
```diff
@@ -1,19 +1,23 @@
...
-#   hlint:
-#     ghc-lib: true
-#   ghc-lib-parser-ex:
-#     auto: false
-#     no-ghc-lib: false
+#  hlint:
+#    ghc-lib: true
+#  ghc-lib-parser-ex:
+#    auto: false
+#    no-ghc-lib: false
```
Comment (unclechu) [inline stack.yaml:23]:
```suggestion
#   hlint:
#     ghc-lib: true
#   ghc-lib-parser-ex:
#     auto: false
#     no-ghc-lib: false
```

File: src/Main.hs [hunk 4]
```diff
@@ -8,7 +8,7 @@ import System.Exit
...
 main :: IO ()
 main = do
-    args <- getArgs
-    errs <- hlint args
-    unless (null errs) $
-        exitWith $ ExitFailure 1
+     args <- getArgs
+     errs <- hlint args
+     unless (null errs) $
+         exitWith $ ExitFailure 1
```
Comment (unclechu) [inline src/Main.hs:14]: 5 spaces for indentation? But still the `exitWith` has 4.

```suggestion
    args <- getArgs
    errs <- hlint args
    unless (null errs) $
        exitWith $ ExitFailure 1
```

File: src/Hint/Restrict.hs [hunk 5]
```diff
@@ -181,14 +180,15 @@ importListToIdents =
...
     fromName :: LIEWrappedName (IdP GhcPs) -> Maybe String
-    fromName wrapped = case unLoc wrapped of
-                         IEName    n -> fromId (unLoc n)
-                         IEPattern n -> ("pattern " ++) <$> fromId (unLoc n)
-                         IEType    n -> ("type " ++) <$> fromId (unLoc n)
+    fromName wrapped =
+      case unLoc wrapped of
+        IEName      n   -> fromId (unLoc n)
+        IEPattern _ n -> ("pattern " ++) <$> fromId (unLoc n)
+        IEType    _ n-> ("type " ++) <$> fromId (unLoc n)
```
Comment (unclechu) [inline src/Hint/Restrict.hs:191]: Broken alignment:
```suggestion
        IEName      n -> fromId (unLoc n)
        IEPattern _ n -> ("pattern " ++) <$> fromId (unLoc n)
        IEType    _ n -> ("type " ++) <$> fromId (unLoc n)
```

File: src/GHC/Util/Scope.hs [hunk 6]
```diff
@@ -38,11 +38,13 @@ scopeCreate xs = Scope $ [prelude | not $ any isPrelude res] ++ res
 
     -- The import declaraions contained by the module 'xs'.
     res :: [LImportDecl GhcPs]
-    res = [x | x <- hsmodImports xs , pkg x /= Just (StringLiteral NoSourceText (fsLit "hint"))]
+    res = [x | x <- hsmodImports xs
+              , pkg x /= Just (StringLiteral NoSourceText (fsLit "hint") Nothing)
+            ]
```
Comment (unclechu) [inline src/GHC/Util/Scope.hs:43]:
```suggestion
    res = [x | x <- hsmodImports xs, pkg x /= Just (StringLiteral NoSourceText (fsLit "hint") Nothing)]
```

**Comment by ndmitchell** [comment]
It's got to the stage where we need to move, so landing, and will fix forward if we need to.

**Comment by ndmitchell** [comment]
Merged, but all the CI is red on GHC 9.2 because apply-refact is not yet GHC 9.2 compatiblehttps://github.com/mpickering/apply-refact/issues/114

**Comment by alanz** [comment]
I am working on the failing tests for apply-refact, and came across `Bracket7`, which has

```haskell
yes = if (f x) then y else z
```
and a corresponding apply-refact refactoring of
```haskell
[("tests/examples/Bracket7.hs:1:10-14: Suggestion: Redundant bracket\nFound:\n  if (f x) then y else z\nPerhaps:\n  if f x then y else z\n"
,[Replace
  { rtype = Expr
  , pos = SrcSpan {startLine = 1, startCol = 10, endLine = 1, endCol = 15}
  , subts = [("x",SrcSpan {startLine = 1, startCol = 11, endLine = 1, endCol = 14})]
  , orig = "x"}])]
```
I presume that only the positions are meaningful, as in "substitute 1:10-15 with 1:11-14", and the reference to `"x"` can be ignored?

**Comment by ndmitchell** [comment]
@alanz  - @zliu41 is the expert here. I think (but am most definitely not sure) that the code is saying:

* Due to `pos` Operate on 1:10-15, i.e. `(f x)`
* Due to `subts` bind `$x` to 1:11-14, i.e. `f x`
* Replace the whole thing with `orig` definition of `$x` so replace `(f x)` with `f x`

But that's just a guess on what I think it's doing.

**Comment by alanz** [comment]
That seems to make sense. So the "x" is a "hlint variable", rather than a quote of existing source code.

**Comment by zliu41** [comment]
@alanz Yeah "x" is a template variable. In this case it happens to be the entire template (`orig`), but in general that's not the case.

**Comment by alanz** [comment]
https://github.com/mpickering/apply-refact/pull/116 is ready for review, should hopefully unblock this

**Pull request merged**